### ISSUE TITLE
Update Spring Security CSRF integration

### ISSUE THREAD

**Issue by dogeared** [stormpath/stormpath-sdk-java#339]
per @lhazlewood:

`<input type="hidden" th:name="${_csrf.parameterName}" th:value="${_csrf.token}"/>`

should work as it's idiomatic Spring Security. Currently, it does not work.

We create a Stormpath `CsrfTokenManager` and then delegate to Spring Security's `CsrfTokenRepository`.

We manually construct an instance of that repository however Spring Boot Security will use the same `CsrfTokenManager` instance for its own needs because of `@ConditionalOnMissingBean` logic.
i.e. Spring Security says "oh, someone already created an instance of `CsrfTokenRepository` - I'll just use that instead of creating one"

Our instance renames both `sessionAttributeName` and `parameterName` to `csrfToken` not `_csrf` and the template fails to render.

It does successfully render if you specify: `<input type="hidden" th:name="${csrfToken.parameterName}" th:value="${csrfToken.token}"/>` since it can successfully find a `csrfToken` bean in the session but a Spring Security user using our integration would never know this and it would break things immediately for them.

**Comment by rwinch** [comment]
@dogeared While this does appear to be an issue, I'm curious why you need to manually specify the CSRF in the first place? Spring Security, Spring MVC, and Thymeleaf provide integration in which the CSRF token should [automatically be added](http://docs.spring.io/spring-security/site/docs/current/reference/html/csrf.html#csrf-include-csrf-token-form).

> If you are using Spring MVC `<form:form>` tag or Thymeleaf 2.1+ and are using `@EnableWebSecurity`, the CsrfToken is automatically included for you (using the `CsrfRequestDataValueProcessor`).

**Comment by lhazlewood** [comment]
@rwinch it's because we have our own lightweight Form/MVC classes that we use in our Servlet integration (when Spring is not in play at all).  When running in a Spring environment, our Spring MVC stuff 'wraps' the existing Stormpath Controllers and acts a bridge from Spring to our Controller API:

https://github.com/stormpath/stormpath-sdk-java/blob/stormpath-sdk-root-1.0.RC5.1/extensions/spring/stormpath-spring-webmvc/src/main/java/com/stormpath/spring/mvc/SpringController.java

Our underlying controller implementation requires CSRF tokens to function correctly.

This ticket refers to the work to ensure it works seamlessly when Spring and Spring Boot are in play, without the user having to configure anything.  The idea is that the same code/functionality works in all web deployments, even if a Stormpath customer may not be using Spring.

HTH!

**Comment by lhazlewood** [comment]
p.s. this was mostly a bridge concept to support the existing functionality - we may find that writing native Spring MVC controllers in the future is worth the effort of maintaining two different sets of controllers.  If we can swing the existing approach, it'd be ideal since we have one set of classes to maintain, but maybe more importantly, use for regression testing to guarantee identical behavior across all web runtime environments.

**Comment by rwinch** [comment]
@lhazlewood Thanks for the fast reply. Perhaps I'm still missing something, but to my understanding it should not matter what your controllers are (it should work with the bridged controllers). Thymeleaf should automatically add the CSRF to forms as long as you have Thymeleaf's thymeleaf-spring4 on the classpath and the form action is using `th:action`.

**Comment by dogeared** [comment]
@lhazlewood - just want to point out here that if you are using the Stormpath Spring Security integration (in Spring Boot or plain Spring form) AND you are using a form that hooks into the Thymeleaf `CsrfRequestDataValueProcessor`, it will put the hidden csrf field in for you. The simplest form of this is:

```
<form th:action="@{/}" method="post">
    <input type="submit"/>
</form>
```

The resulting html source is:

```
<form method="post" action="/">
    <input type="hidden" name="csrfToken" value="0ae820bb-0115-4234-8368-e61e7136d652" />
    <input type="submit" />
</form>
```

You only run into the problem this issue is about if you explicitly try to use the ${_csrf...} stuff as we don't have that.

**Comment by dogeared** [comment]
Addressed by #348
